helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -31,69 +31,162 @@
    * @param lockPath the path of the lock under Zookeeper
-   * @param timeout the timeout period of the lock
+   * @param leaseTimeout the leasing timeout period of the lock
    * @param lockMsg the reason for having this lock
    * @param userId a universal unique userId for lock owner identity
+   * @param priority the priority of the lock
+   * @param waitingTimeout the waiting timeout period of the lock when the tryLock request is issued
+   * @param cleanupTimeout the time period needed to finish the cleanup work by the lock when it
+   *                      is preempted
+   * @param isForceful whether the lock is a forceful one. This determines the behavior when the
+   *                   lock encountered an exception during preempting lower priority lock
+   * @param lockListener the listener associated to the lock
    * @param baseDataAccessor baseDataAccessor instance to do I/O against ZK with
    */
-  private ZKDistributedNonblockingLock(String lockPath, Long timeout, String lockMsg, String userId,
-      BaseDataAccessor<ZNRecord> baseDataAccessor) {
+  private ZKDistributedNonblockingLock(String lockPath, Long leaseTimeout, String lockMsg,
+      String userId, int priority, long waitingTimeout, long cleanupTimeout, boolean isForceful,
+      LockListener lockListener, BaseDataAccessor<ZNRecord> baseDataAccessor) {
     _lockPath = lockPath;
-    if (timeout < 0) {
-      throw new IllegalArgumentException("The expiration time cannot be negative.");
+    if (leaseTimeout < 0 || waitingTimeout < 0 || cleanupTimeout < 0) {
+      throw new IllegalArgumentException("Timeout cannot be negative.");
+    }
+    if (priority < 0) {
+      throw new IllegalArgumentException("Priority cannot be negative.");
     }
-    _timeout = timeout;
+    _leaseTimeout = leaseTimeout;
     _lockMsg = lockMsg;
     _userId = userId;
     _baseDataAccessor = baseDataAccessor;
+    _priority = priority;
+    _waitingTimeout = waitingTimeout;
+    _cleanupTimeout = cleanupTimeout;
+    _lockListener = lockListener;
+    _isForceful = isForceful;
   }
 
   @Override
-  public boolean tryLock() {
+  public synchronized boolean tryLock() {
     // Set lock information fields
-    long deadline;
-    // Prevent value overflow
-    if (_timeout > Long.MAX_VALUE - System.currentTimeMillis()) {
-      deadline = Long.MAX_VALUE;
-    } else {
-      deadline = System.currentTimeMillis() + _timeout;
+    _baseDataAccessor.subscribeDataChanges(_lockPath, this);
+    LockUpdater updater = new LockUpdater(
+        new LockInfo(_userId, _lockMsg, getNonOverflowTimestamp(_leaseTimeout), _priority,
+            _waitingTimeout, _cleanupTimeout, null, 0, 0, 0));
+    boolean updateResult = _baseDataAccessor.update(_lockPath, updater, AccessOption.PERSISTENT);
+
+    // Check whether the lock request is still pending. If yes, we will wait for the period
+    // recorded in _pendingTimeout.
+    if (_isPending) {
```

Comment:
Usually, if we are waiting for something, we should use while to do the checks here.

Reply:
This boolean variable is used as a flag to distinguish between different scenarios so that we can act differently. The actual functioning part is the following wait().

Reply:
The reason to put wait in the while loop is to prevent if anyone else updates the content after the notification and before the current thread regains the lock. I think what Junkai said is a valid comment.

Reply:
If there's another user (whose priority is even higher) acquired the lock, the onDeniedPendingLockNotification will be called, which will terminate the wait. We won't need to worry about this, right?

Reply:
I see your point. So here we cannot use while because pending is not the only checking condition. We also have timeout, which can terminate the wait. We cannot easily distinguish between spurious wake up and timeout. Using a timer is not accurate. I changed to use countdown latch, which is more convenient and doesn't have spurious wake up issue.